fvm/crypto/crypto.go

```diff
@@ -160,71 +164,85 @@ func VerifySignatureFromRuntime(
 		// there are no tag constraints
 	}
 
+	// decode the public key
 	publicKey, err := crypto.DecodePublicKey(sigAlgo, rawPublicKey)
 	if err != nil {
 		return false, errors.NewValueErrorf(hex.EncodeToString(rawPublicKey), "cannot decode public key: %w", err)
 	}
 
-	valid, err := verifier.Verify(
-		signature,
-		tag,
-		message,
-		publicKey,
-		hashAlgo,
-	)
+	// create a hasher
+	var hasher hash.Hasher
+	switch hashAlgo {
+	case hash.SHA2_256, hash.SHA3_256, hash.Keccak_256:
+		var err error
+		if hasher, err = NewPrefixedHashing(hashAlgo, tag); err != nil {
+			return false, errors.NewValueErrorf(err.Error(), "runtime verification failed")
+		}
+	case hash.KMAC128:
+		hasher = crypto.NewBLSKMAC(tag)
+	default:
+		return false, errors.NewValueErrorf(fmt.Sprint(hashAlgo), "hashing algorithm type not found")
+	}
+
+	valid, err := publicKey.Verify(signature, message, hasher)
 	if err != nil {
-		return false, err
+		// All inputs are guaranteed to be valid at this stage.
+		// The check for crypto.InvalidInputs is only a sanity check
+		if crypto.IsInvalidInputsError(err) {
+			return false, err
+		}
+		panic(fmt.Errorf("verify runtime signature failed with unexpected error %w", err))
 	}
 
 	return valid, nil
 }
 
-type SignatureVerifier interface {
-	Verify(
-		signature []byte,
-		tag string,
-		message []byte,
-		publicKey crypto.PublicKey,
-		hashAlgo hash.HashingAlgorithm,
-	) (bool, error)
-}
-
-type DefaultSignatureVerifier struct{}
-
-func NewDefaultSignatureVerifier() DefaultSignatureVerifier {
-	return DefaultSignatureVerifier{}
-}
-
-func (DefaultSignatureVerifier) Verify(
+// VerifySignatureFromRuntime performs signature verification using raw values provided
+// by the Cadence runtime.
+//
+// The signature/hash function combinations accepted are:
+//   - ECDSA (on both curves P-256 and secp256k1) with any of SHA2-256/SHA3-256.
+// The tag is applied to the message as a constant length prefix.
+//
+// The function errors:
+//  - NewValueErrorf for any user error
+//  - panic for any other unexpected error
+func VerifySignatureFromTransaction(
 	signature []byte,
-	tag string,
 	message []byte,
-	publicKey crypto.PublicKey,
+	pk crypto.PublicKey,
 	hashAlgo hash.HashingAlgorithm,
 ) (bool, error) {
 
-	var hasher hash.Hasher
+	// check ECDSA compatibilites
+	if pk.Algorithm() != crypto.ECDSAP256 && pk.Algorithm() != crypto.ECDSASecp256k1 {
+		// TODO: check if we should panic
+		// This case only happens in production if there is a bug
+		return false, errors.NewUnknownFailure(fmt.Errorf(
+			pk.Algorithm().String(), "is not supported in transactions"))
+	}
+	// hashing compatibility
+	if hashAlgo != hash.SHA2_256 && hashAlgo != hash.SHA3_256 {
+		// TODO: check if we should panic
+		// This case only happens in production if there is a bug
+		return false, errors.NewUnknownFailure(fmt.Errorf(
+			hashAlgo.String(), "is not supported in transactions"))
+	}
```

Comment:
not sure about this, you mean if there is some problematic old keys we would stop the network? I think we should just error and add a watch on the error type.

Reply:
I tagged you about this [here](https://github.com/onflow/flow-go/pull/2171#discussion_r832795320). I suggested to panic, Janez suggested the Failure error. As mentioned in the comment, that case can only happen if there is a bug in the code. These keys are all account keys, they are supposed to be checked and added previously, and they are not supposed to be problematic. 

What do you suggest to use here?